[bitnami/common] fix failOnNew implementation in common.secrets.passwords.manage

fty4 opened the pull request:
### Description of the change

Using a string validation for the `failOnNew` param instead of the current boolean implementation (not working).
Also patching kafka chart where the param is currently used.

### Benefits

This would allow to use the `failOnNew` param as intended.
Currently it does not effect if the variable is changed or not.

### Possible drawbacks


### Applicable issues

- fixes #21304

### Additional information

You will find a detailed description why this should be fixed in the related [issue](https://github.com/bitnami/charts/issues/21304).

Not sure how to name the title of the PR - `[bitnami/common]` would also fit but updated kafka chart as well.

### Checklist

<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

<hr />

<sub>Marco Lecheler [marco.lecheler@mercedes-benz.com](mailto:marco.lecheler@mercedes-benz.com) Mercedes-Benz Tech Innovation GmbH ([ProviderInformation](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md))</sub>

juan131 reviewed (changes requested):
Hi @fty4 !! Thanks so much for this contribution, could you please move the changes in Kafka to a different PR? The right order to apply these changes would be:

- Merging the PR containing the changes in "common" chart
- Waiting for CI/CD to release the new "common" chart version
- Update deps (`Chart.lock`) on Kafka to use the new "common" chart version and apply the required changes

fty4 commented:
Hi @juan131 I've just removed the commits from this PR as requested - please have a look.

juan131 commented on bitnami/common/templates/_secrets.tpl:
```diff
@@ -91,7 +91,7 @@ The order in which this function returns a secret password:
 {{- $password := "" }}
 {{- $subchart := "" }}
-{{- $failOnNew := default true .failOnNew }}
+{{- $failOnNew := default "true" .failOnNew }}
```
I just saw that you forgot to update "Usage" information (see line 80) defining:
```diff
-  - failOnNew - Boolean - Optional - Default to true. If set to false, skip errors adding new keys to existing secrets.
+  - failOnNew - String - Optional - Default to "true". If set to "false", skip errors adding new keys to existing secrets.
```

juan131 commented on bitnami/common/templates/_secrets.tpl:
```diff
@@ -100,7 +100,7 @@ The order in which this function returns a secret password:
   {{- if hasKey $secretData .key }}
     {{- $password = index $secretData .key | quote }}
-  {{- else if $failOnNew }}
+  {{- else if eq $failOnNew "true" }}
```
Before discarding using booleans, could you try this?

- Remove use or `default` to set `$failOnNew`:

```diff
-{{- $failOnNew := default "true" .failOnNew }}
```

- Use `if not` here:

```suggestion
  {{- else if not .failOnNew }}
```

Then, keep using a boolean on Kafka chart.

fty4 replied:
Wouldn't that mean the `failOnNew` can not be set to false?

Also in my test I was not able to use `failOnNew` without Helm asking to provide the current password via additional `--set` parameter.

Could you verify if it is working for you?

fty4 commented:
Sure - that testing works but I think we need to test that a little bit different here.
Because the param `failOnNew` will only apply if the secret already exists.
It needs to be tested if a new key is added (see its description: _skip errors adding new keys to existing secrets_).
Your upgrade did not add a new key.

Therefore I think the test should be as follows:

1. install the Helm chart without changes (expect the patched common dependency)
2. add a new secret to the kafka chart:
```diff
diff --git a/bitnami/kafka/templates/secrets.yaml b/bitnami/kafka/templates/secrets.yaml
index 7243ee417..8d0838c74 100644
--- a/bitnami/kafka/templates/secrets.yaml
+++ b/bitnami/kafka/templates/secrets.yaml
@@ -16,6 +16,8 @@ metadata:
   {{- end }}
 type: Opaque
 data:
+  inter-broker-password-test: {{ include "common.secrets.passwords.manage" (dict "secret" $secretName "key" "inter-broker-password-test" "providedValues" (list "sasl.interbroker.password") "context" $) }}
   {{- if (include "kafka.client.saslEnabled" .) }}
   {{- $secretValue := "" }}
   {{- $secretData := (lookup "v1" "Secret" (include "common.names.namespace" .) $secretName).data }}
```
3. Upgrade the chart

## Tests

I've tested both changes (current PR and your suggestions) as follows:

1. Current PR change without your changes and `"failOnNew" "false"`
Helm update successful ✅

2. Current PR change without your changes and `"failOnNew" "true"`
```
Helm upgrade failed ✅
$ helm upgrade kafka ./bitnami/kafka -f ./values.yaml
Error: UPGRADE FAILED: execution error at (kafka/templates/controller-eligible/statefulset.yaml:42:38): 
PASSWORDS ERROR: The secret "kafka-user-passwords" does not contain the key "inter-broker-password-test"
```

3. Current PR change without your changes and `failOnNew` not specified
```
Helm upgrade failed ✅
$ helm upgrade kafka ./bitnami/kafka -f ./values.yaml
Error: UPGRADE FAILED: execution error at (kafka/templates/controller-eligible/statefulset.yaml:42:38): 
PASSWORDS ERROR: The secret "kafka-user-passwords" does not contain the key "inter-broker-password-test"
```

4. Current PR change with your changes and `"failOnNew" false`
```
Helm upgrade failed ❌
$ helm upgrade kafka ./bitnami/kafka -f ./values.yaml
Error: UPGRADE FAILED: execution error at (kafka/templates/controller-eligible/statefulset.yaml:42:38): 
PASSWORDS ERROR: The secret "kafka-user-passwords" does not contain the key "inter-broker-password-test"
```

5. Current PR change with your changes and `"failOnNew" true`
```
Helm update failed (failing ok but wrong error) ❌
$ helm upgrade kafka ./bitnami/kafka -f ./values.yaml
Error: UPGRADE FAILED: template: kafka/templates/controller-eligible/statefulset.yaml:42:38: executing "kafka/templates/controller-eligible/statefulset.yaml" at <include (print $.Template.BasePath "/secrets.yaml") .>: error calling include: template: kafka/templates/secrets.yaml:19:33: executing "kafka/templates/secrets.yaml" at <include "common.secrets.passwords.manage" (dict "secret" $secretName "key" "inter-broker-password-test" "providedValues" (list "sasl.interbroker.password") "failOnNew" true "context" $)>: error calling include: template: kafka/charts/common/templates/_secrets.tpl:103:14: executing "common.secrets.passwords.manage" at <eq $failOnNew "true">: error calling eq: incompatible types for comparison
```

6. Current PR change with your changes and `failOnNew` not specified
```
Helm upgrade failed ✅
$ helm upgrade kafka ./bitnami/kafka -f ./values.yaml
Error: UPGRADE FAILED: execution error at (kafka/templates/controller-eligible/statefulset.yaml:42:38): 
PASSWORDS ERROR: The secret "kafka-user-passwords" does not contain the key "inter-broker-password-test"
```

That would mean the additional changes will not behave as expected, right?

juan131 commented:
Thanks for the details @fty4 

> Your upgrade did not add a new key

It did, please note `controller-password` wasn't present in the original secret. 

> I've tested both changes (current PR and your suggestions) as follows:
> (...)

I'm unable to reproduce the failures steps 4 & 5. Did you locally package the `common` chart with the changes and replace `common` dep chart within Kafka dependencies before running the tests?

fty4 commented:
I am sorry.. shame on me: I forgot to update the dependencies after updating the `_secrets.tpl`.
Your changes are working as desired - thank you for your feedback @juan131.

Just added the changes to the PR.
Also reverted the param change back to boolean.

_Required force-push because of missing sign-off commits_

juan131 reviewed (approved):
LGTM

juan131 commented:
Thanks so much for your contribution!
